src/Language/DifferentialDatalog/rust/datalog/program.rs

```diff
@@ -0,0 +1,856 @@
+                            for update in updates.drain(..) {
+                                match update {
+                                    Update::Insert{relid, v} => {
+                                        sessions.get_mut(&relid).unwrap().insert(v);
+                                    },
+                                    Update::Delete{relid, v} => {
+                                        sessions.get_mut(&relid).unwrap().remove(v);
+                                    }
+                                }
+                            };
+                            epoch = epoch+1;
+                            //print!("epoch: {}\n", epoch);
+                            Self::advance(&mut sessions, epoch);
+                        },
+                        Ok(Msg::Flush) => {
+                            //println!("flushing");
+                            Self::flush(&mut sessions, &probe, worker);
+                            //println!("flushed");
+                            flush_ack_send.send(()).unwrap();
+                        },
+                        Ok(Msg::Stop) => {
+                            break;
+                        }
+                    };
+                };
+            }));
+
+        //println!("timely computation started");
+
+        let mut rels = FnvHashMap::default();
+        for relid in self.input_relations() {
+            rels.insert(relid, 
+                        RelationInstance{
+                            input:    self.get_relation(relid).input,
+                            elements: FnvHashSet::default(),
+                            delta:    FnvHashMap::default()
+                        });
+        };
+
+        RunningProgram{
+            sender: tx,
+            flush_ack: flush_ack_recv,
+            relations: rels,
+            thread_handle: h,
+            transaction_in_progress: false,
+            need_to_flush: false
+        }
+    }
+
+    /* Advance the epoch on all input sessions */
+    fn advance(sessions: &mut FnvHashMap<RelId, InputSession<u64, V, isize>>, epoch : u64) {
+        for (_,s) in sessions.into_iter() {
+            //print!("advance\n");
+            s.advance_to(epoch);
+        };
+    }
+
+    /* Propagate all changes through the pipeline */
+    fn flush(
+        sessions: &mut FnvHashMap<RelId, InputSession<u64, V, isize>>,
+        probe: &ProbeHandle<Product<RootTimestamp, u64>>,
+        worker: &mut Root<Allocator>) 
+    {
+        for (_,r) in sessions.into_iter() {
+            //print!("flush\n");
+            r.flush();
+        };
+        if let Some((_,session)) = sessions.into_iter().nth(0) {
+            while probe.less_than(session.time()) {
+                //println!("flush.step");
+                worker.step();
+            };
+        }
+    }
+
+    /* Lookup relation by id */
+    fn get_relation(&self, relid: RelId) -> &Relation<V> {
+        for node in &self.nodes {
```

Comment:
linear scan?

Reply:
yes, but never used in the critical path